proposals/p5337.md

```diff
@@ -0,0 +1,1057 @@
+# Interface extension and `final impl` update
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/5337)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+    -   [`extend require`](#extend-require)
+    -   [`extend impl as` with parameterized interfaces](#extend-impl-as-with-parameterized-interfaces)
+    -   [`extend impl as` restrictions](#extend-impl-as-restrictions)
+    -   [Name conflicts](#name-conflicts)
+    -   [Final `impl` priority](#final-impl-priority)
+        -   [Overlapping `final` impls](#overlapping-final-impls)
+        -   [`impl` selection algorithm](#impl-selection-algorithm)
+        -   [An `impl` that can never match is an error](#an-impl-that-can-never-match-is-an-error)
+    -   [Impl names for prioritization](#impl-names-for-prioritization)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [Allow overlap between a non-final and final impl but only if no queries pick the non-final on the overlap](#allow-overlap-between-a-non-final-and-final-impl-but-only-if-no-queries-pick-the-non-final-on-the-overlap)
+    -   [Forbid overlap between final impls](#forbid-overlap-between-final-impls)
+    -   [Prioritize between final impls using type structure](#prioritize-between-final-impls-using-type-structure)
+    -   [Allow mixing final and non-final impls in the same `match_first` block](#allow-mixing-final-and-non-final-impls-in-the-same-match_first-block)
+    -   [No default `Self` in `require Self impls I`](#no-default-self-in-require-self-impls-i)
+
+<!-- tocstop -->
+
+## Abstract
+
+We make 5 changes:
+
+-   Allow `require Self impls I` in an `interface` or `constraint` scope to omit
+    the `Self`, so it can be written `require impls I`.
+-   Rename `extend I` to `extend require impls I` in an `interface` or
+    `constraint` scope.
+-   Define `extend impl as I` and `extend final impl as I` in an `interface`
+    scope to copy the members of `I` and define an `impl` of `I` in terms of the
+    extending interface.
+-   Allow a non-final `impl` to overlap a final `impl` as long as it isn't
+    subsumed by the final `impl`. The final `impl` will be given priority on the
+    overlap.
+-   Allow `final` on a `match_first` block, used to declare overlapping final
```

Comment:
This does not mention that final impls can overlap if one comes from the root self type library and one from the interface library (and they are different libraries). Either this should be generalized or made to include both cases?